src/java.base/share/classes/com/sun/crypto/provider/SHA3Parallel.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package sun.security.provider;
+
+import jdk.internal.vm.annotation.IntrinsicCandidate;
+
+import java.util.Arrays;
+
+import static sun.security.provider.ByteArrayAccess.b2lLittle;
+import static sun.security.provider.ByteArrayAccess.l2bLittle;
+
+import static sun.security.provider.SHA3.keccak;
+
+public class SHA3Parallel {
```

Comment:
Why not merge this with `sun.security.provider.SHA3` class? A separate class in a different package seems harder to track...

Reply:
@ferakocz might have some more insight here

Reply:
It is in the same package. It is not a full implementation of doing two SHA3 XOF operations in parallel, it is assumed that the two state arrays that are passed in at instantiation/reset contain the state after the absorb phase of the XOF, where the length of the whole input was less than one block size, and the parallel part only does the squeezes, one pair of blocks per call. So it is better to keep it in a separate file, although it should definitely be moved to the correct place in the source file hierarchy (src/java.base/share/classes/sun/security/provider/SHA3Parallel.java).